Add wsl-proxy forwarder

Nino-K opened the pull request:
This change introduces the wsl proxy forwarder, the forwarder is responsible for forwarding all the port binding events from docker, containerd, and kubernetes over the UNIX socket to wsl-proxy process that runs outside of the namespace in the default namespace. The wsl proxy forwarder implements a forwarder interface so that it is easily integrated able into the corresponding tracker, e.g. apitracker (which is used for the network namespace).

Related: https://github.com/rancher-sandbox/rancher-desktop/issues/6636

mook-as commented on pkg/forwarder/wslproxy.go:
```diff
@@ -0,0 +1,51 @@
+*/
+
+// Package forwarder implements a forwarding mechanism to forward
+// port mappings to WSL Proxy.
```
Needs more information, especially since "WSL Proxy" sounds very generic. When is this used? (I guess forwarding from the namespaced network to the root network namespace inside of WSL?) Link to the WSL proxy repo?

mook-as commented on pkg/tracker/apitracker.go:
```diff
@@ -46,13 +48,15 @@ var (
 	ErrExposeAPI   = fmt.Errorf("error from %s API", exposeAPI)
 	ErrUnexposeAPI = fmt.Errorf("error from %s API", unexposeAPI)
 	ErrInvalidIPv4 = errors.New("not an IPv4 address")
+	ErrWSLProxy    = errors.New("error from WSL Proxy")
```
In context, "WSL Proxy" is a terrible name (there are too many things it can be); this message needs to be clearer about _which_ WSL proxy it's talking about.

mook-as reviewed (changes requested):
The commit needs to be split up so that the relevant changes can be reviewed separately from the mechanical changes.

Nino-K committed:
Rename testVTunnelForwarder to testForwarder

So it can be used as a generic forwarder in
apitracker tests.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Add a forwarder to API Tracker

So the API tracker is able to forward the port mappings
to rancher desktop wsl-proxy.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Add WSLProxy forwarder

The WSLProxyForwarder allows to forward port mappings to
the rancher desktop wsl proxy over the unix socket.

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

Nino-K committed:
Update go mod

Signed-off-by: Nino Kodabande <nkodabande@suse.com>

mook-as commented on pkg/forwarder/wslproxy.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+/*
+This package provides functionality to forward port mappings to the
```
If this is mean to be the package-level documentation, that should go in front of the `package forwarder` line, right? (But we have something there already)

Nino-K replied:
Fixed.

mook-as commented on pkg/forwarder/wslproxy.go:
```diff
@@ -0,0 +1,59 @@
+
+// Send forwards the port mappings to WSL Proxy.
+func (v *WSLProxyForwarder) Send(portMapping types.PortMapping) error {
+	conn, err := net.Dial("unix", v.proxySocket)
```
I assume we don't dial in the constructor to avoid issues with figuring out which process starts first? And it's not worth the effort to cache the connection and reuse it for subsequent calls?

(I'm not questioning the decisions, just to check my assumptions; it seems perfectly fine as it is.)

Nino-K replied:
Yeah, that's correct. Also, all these forwarders operate in a fire-and-forget manner, basically, a new connection is established the port mappings are forwarded and then the connection is closed immediately.

mook-as reviewed (approved):
Just some boring process questions; can merge once you read them.

Nino-K committed:
Fixes the package docs for wslproxy

Signed-off-by: Nino Kodabande <nkodabande@suse.com>
